doc/guides/build/index.rst

```diff
@@ -3,448 +3,10 @@
-
-.. include:: ../../../scripts/gen_kobject_list.py
-   :start-after: """
-   :end-before: """
-
-.. _gen_offset_header.py:
-
-:zephyr_file:`scripts/gen_offset_header.py`
--------------------------------------------
-
-.. include:: ../../../scripts/gen_offset_header.py
-   :start-after: """
-   :end-before: """
-
-.. _parse_syscalls.py:
-
-:zephyr_file:`scripts/parse_syscalls.py`
-----------------------------------------
-
-
-.. include:: ../../../scripts/parse_syscalls.py
-   :start-after: """
-   :end-before: """
-
-.. _gen_idt.py:
-
-:zephyr_file:`arch/x86/gen_idt.py`
-----------------------------------
-
-.. include:: ../../../arch/x86/gen_idt.py
-   :start-after: """
-   :end-before: """
-
-.. _gen_gdt.py:
-
-:zephyr_file:`arch/x86/gen_gdt.py`
-----------------------------------
-
-.. include:: ../../../arch/x86/gen_gdt.py
-   :start-after: """
-   :end-before: """
-
-.. _gen_relocate_app.py:
-
-:zephyr_file:`scripts/gen_relocate_app.py`
--------------------------------------------
-
-.. include:: ../../../scripts/gen_relocate_app.py
-   :start-after: """
-   :end-before: """
-
-.. _process_gperf.py:
-
-:zephyr_file:`scripts/process_gperf.py`
----------------------------------------
-
-.. include:: ../../../scripts/process_gperf.py
-   :start-after: """
-   :end-before: """
-
-:zephyr_file:`scripts/gen_app_partitions.py`
---------------------------------------------
-
-.. include:: ../../../scripts/gen_app_partitions.py
-   :start-after: """
-   :end-before: """
-
-.. _kconfig:
-
-Configuration System (Kconfig)
-*******************************
-
-The Zephyr kernel and subsystems can be configured at build time to adapt them
-for specific application and platform needs. Configuration is handled through
-Kconfig, which is the same configuration system used by the Linux kernel. The
-goal is to support configuration without having to change any source code.
-
-Configuration options (often called *symbols*) are defined in :file:`Kconfig`
-files, which also specify dependencies between symbols that determine what
-configurations are valid. Symbols can be grouped into menus and sub-menus to
-keep the interactive configuration interfaces organized.
-
-The output from Kconfig is a header file :file:`autoconf.h` with macros that
-can be tested at build time. Code for unused features can be compiled out to
-save space.
-
-The following sections explain how to set Kconfig configuration options, go
-into detail on how Kconfig is used within the Zephyr project, and have some
-tips and best practices for writing :file:`Kconfig` files.
-
 .. toctree::
-   :maxdepth: 1
-
-   kconfig/menuconfig.rst
-   kconfig/setting.rst
-   kconfig/tips.rst
-   kconfig/preprocessor-functions.rst
-   kconfig/extensions.rst
+   :maxdepth: 2
 
-Users interested in optimizing their configuraion for security should refer
-to the Zephyr Security Guide's section on the :ref:`hardening`.
+   build.rst
+   dts/index.rst
+   kconfig/index.rst
+   dt-vs-kconfig.rst
```

Comment:
Instead of having this DT vs Kconfig as a top level item, it would be better to fold this either into DTS or Kconfig as an intro section.  This way we have a clean structure with no loose items at this index level.

Reply:
I don't agree. This is a page distinguishing the two of them, it does not logically belong in one or the other, since they serve different purposes.

Reply:
Can you define "loose item" and explain why it is a problem? I really see no issue here.

Reply:
>  I don't agree. This is a page distinguishing the two of them, it does not logically belong in one or the other, since they serve different purposes.

You can also have two sections: 
 - Build System
 - Configuration System
    -  Device Tree vs Kconfig
    -  Kconfig
    -  Device Tree

Reply:
> Can you define "loose item" and explain why it is a problem? I really see no issue here.

If we start adding items at this level, one more page regarding build system and another about something else talking about configuration system, you lose the structure.

Reply:
> you lose the structure.

I don't agree.

The top level toctree is for build and configuration systems -- systems plural, since DT and Kconfig are distinct configuration systems:

1. build system
2. first configuration system, kconfig
3. second configuration  system B (DT)

We have all agreed that this makes sense over on Discord.

Now, we also definitely want some information comparing and contrasting the two configuration systems.

You are saying we should have:

1. build system
2. config system A
3. config system B
     3.1. config system A vs config system B

I am saying we should have:

1. build system
2. config system A
3. config system B
4. config system A vs config system B

In my opinion, this is a better structure: you want the section comparing A and B at the same level as them; it doesn't make sense to hide it underneath one of them.

Looking back at the history of the project, I think the only reason the existing documentation has "kconfig versus devicetree" in a sub-heading is because devicetree came later to the project than kconfig. But that's ancient history now; let's structure this properly.

@nashif please revisit your review.

@carlescufi please offer feedback as documentation maintainer.

Reply:
> You are saying we should have:
> 
> 1. build system
> 2. config system A
> 3. config system B
>    3.1. config system A vs config system B

I am actually saying:

- Build System
   -  Cmake...
   - Here we can also move the cmake package doc
   -  other build system related docs 
- Configuration Systems
  - Intro from the A vs B
  - Kconfig
  - DTS

This [page](https://docs.zephyrproject.org/latest/guides/dts/dt-vs-kconfig.html)  was added at a time when we were figuring things out and it was needed to clarify the direction we were taking. IMO this needs to be redone and added as the intro for Configuration systems.

Reply:
Well, I am going to go back on the comment [I made](https://github.com/zephyrproject-rtos/zephyr/pull/41154#pullrequestreview-831714519) in @nashif 's PR to state exactly the same here: Let us decide on a top-level structure first before we start moving things around. I suggest we use #33725 to capture the discussion about the top-level hierarchy, then implement it using these PRs.

Reply:
> I am actually saying:
> 
> * Build System
>   
>   * Cmake...
>   * Here we can also move the cmake package doc
>   * other build system related docs
> * Configuration Systems
>   
>   * Intro from the A vs B
>   * Kconfig
>   * DTS

I'd rather we discuss in #33725, but my vote in this case would go for:

- Build and Configuration Systems
  - Build System
     - CMake
     - CMake Package
     - Other
  - Configuration systems
     - Devicetree vs Kconfig
     - Kconfig
     - Devicetree

Reply:
> * Build and Configuration Systems
>   
>   * Build System
>     
>     * CMake
>     * CMake Package
>     * Other
>   * Configuration systems
>     
>     * Devicetree vs Kconfig
>     * Kconfig
>     * Devicetree

yes, that is in line with what I would like to see, I omitted the top level section containing both build and config systems  assuming it was obvious.